dali/core/os/shared_mem.cc

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <fcntl.h> /* For O_* constants */
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <cstring>
+#include <exception>
+
+#include "dali/core/format.h"
+#include "dali/core/os/shared_mem.h"
+
+namespace dali {
+
+namespace detail {
+
+bool dir_exists(const char *str) {
+  struct stat sb;
+  if (stat(str, &sb) == -1) {
+    return false;
+  }
+  return S_ISDIR(sb.st_mode);
+}
+
+}  // namespace detail
+
+
+ShmHandle ShmHandle::CreateHandle() {
+  // Abstract away the fact that shm_open requires filename.
+  constexpr char dev_shm_path[] = "/dev/shm/";
+  constexpr char run_shm_path[] = "/run/shm/";
+  constexpr char temp_filename_template[] = "nvidia_dali_XXXXXX";
+  constexpr int kDevShmPathSize = sizeof(dev_shm_path) - 1;
+  constexpr int kRunShmPathSize = sizeof(run_shm_path) - 1;
+  constexpr int kPathSizeMax =
+      std::max(kDevShmPathSize, kRunShmPathSize) + sizeof(temp_filename_template);
+  char temp_path[kPathSizeMax];
+  const char *shm_path = nullptr;
+  size_t shm_size = 0;
+  if (detail::dir_exists(dev_shm_path)) {
+    shm_path = dev_shm_path;
+    shm_size = kDevShmPathSize;
+  } else if (detail::dir_exists(run_shm_path)) {
+    shm_path = run_shm_path;
+    shm_size = kRunShmPathSize;
+  } else {
+    throw std::runtime_error(make_string("Shared memory dir not found, looked for: {", dev_shm_path,
+                                         ", ", run_shm_path, "}."));
+  }
+  memcpy(temp_path, shm_path, shm_size);
+  memcpy(temp_path + shm_size, temp_filename_template, sizeof(temp_filename_template));
+  int temp_fd = mkstemp(temp_path);  // posix
+  POSIX_CHECK_STATUS_EX(temp_fd, "mkstemp", "Temporary file creation failed.");
+  POSIX_CALL_EX(unlink(temp_path), "Couldn't unlink temporary file.");
+  const char *temp_filename = temp_path + shm_size;
+  shm_handle_t fd = shm_open(temp_filename, O_CREAT | O_EXCL | O_RDWR, S_IRUSR | S_IWUSR);
+  POSIX_CHECK_STATUS(temp_fd, "shm_open");
```

Comment:
```suggestion
  POSIX_CHECK_STATUS(fd, "shm_open");
```
?
In any case, if this fails, `temp_fd` is leaked.

Reply:
wrapped it in the ShmHandle, it should be now closed it the `fd` was invalid (as we will throw).

Reply:
And this was classic copy-paste.